File: src/shared/utils/pathway.ts
import { GridRow, PathwayConfig, PathwayPeriodConfig, Primitive, WorkbookModel } from '../types';

export const PATHWAY_CONFIG_SHEET = 'RAGNAROK_Pathway';
export const PATHWAY_PERIODS_SHEET = 'RAGNAROK_PathwayPeriods';

export function defaultPathwayConfig(): PathwayConfig {
  return {
    planningMode: 'single_period',
    enabled: false,
    snapshotMappingMode: 'explicit_period_column',
    overridePolicy: 'reuse_base_inputs',
    periods: [],
    selectedPeriod: null,
  };
}

function primitiveBoolean(value: Primitive, fallback: boolean): boolean {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number') return value !== 0;
  if (typeof value === 'string') {
    const normalized = value.trim().toLowerCase();
    if (normalized === 'true' || normalized === '1' || normalized === 'yes') return true;
    if (normalized === 'false' || normalized === '0' || normalized === 'no') return false;
  }
  return fallback;
}

function primitiveNumber(value: Primitive, fallback: number | null = null): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string') {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return fallback;
}

export function readPathwayConfigFromModel(model: WorkbookModel): PathwayConfig {
  const configRow = (model[PATHWAY_CONFIG_SHEET] ?? [])[0] ?? {};
  const periods = (model[PATHWAY_PERIODS_SHEET] ?? [])
    .map((row): PathwayPeriodConfig | null => {
      const period = primitiveNumber(row.period as Primitive, null);
      if (period === null) return null;
      return {
        period,
        objectiveWeight: primitiveNumber(row.objectiveWeight as Primitive, 1) ?? 1,
        yearsWeight: primitiveNumber(row.yearsWeight as Primitive, 1) ?? 1,
      };
    })
    .filter((row): row is PathwayPeriodConfig => !!row)
    .sort((left, right) => left.period - right.period);

  const selected = primitiveNumber(configRow.selectedPeriod as Primitive, null);
  const enabled = primitiveBoolean(configRow.enabled as Primitive, periods.length > 0);
  return {
    planningMode: enabled ? 'pathway' : 'single_period',
    enabled,
    snapshotMappingMode:
      configRow.snapshotMappingMode === 'repeat_all_snapshots'
        ? 'repeat_all_snapshots'
        : 'explicit_period_column',
    overridePolicy: 'reuse_base_inputs',
    periods,
    selectedPeriod: selected,
  };
}

export function writePathwayConfigToModel(
  model: WorkbookModel,
  config: PathwayConfig,
): WorkbookModel {
  const configRows: GridRow[] = [{
    enabled: config.enabled,
    planningMode: config.planningMode,
    snapshotMappingMode: config.snapshotMappingMode,
    overridePolicy: config.overridePolicy,
    selectedPeriod: config.selectedPeriod,
  }];
  const periodRows: GridRow[] = config.periods.map((period) => ({
    period: period.period,
    objectiveWeight: period.objectiveWeight,
    yearsWeight: period.yearsWeight,
  }));
  return {
    ...model,
    [PATHWAY_CONFIG_SHEET]: configRows,
    [PATHWAY_PERIODS_SHEET]: periodRows,
  };
}

export function samePathwayConfig(left: PathwayConfig, right: PathwayConfig): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

export function getDefaultSelectedPeriod(config: PathwayConfig): number | null {
  if (!config.periods.length) return null;
  if (config.selectedPeriod !== null && config.periods.some((row) => row.period === config.selectedPeriod)) {
    return config.selectedPeriod;
  }
  return config.periods[0].period;
}

File: backend/lib/pathway.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .utils.coerce import number


@dataclass
class PathwayPeriod:
    period: int
    objective_weight: float
    years_weight: float


@dataclass
class PathwayConfig:
    enabled: bool
    planning_mode: str
    snapshot_mapping_mode: str
    periods: list[PathwayPeriod]
    selected_period: int | None


def parse_pathway_config(raw: dict[str, Any] | None) -> PathwayConfig:
    raw = raw or {}
    enabled = bool(raw.get("enabled")) or str(raw.get("planningMode")) == "pathway"
    periods: list[PathwayPeriod] = []
    for item in raw.get("periods") or []:
      try:
        period = int(number(item.get("period")))
      except Exception:
        continue
      periods.append(
        PathwayPeriod(
          period=period,
          objective_weight=float(number(item.get("objectiveWeight"), 1.0)),
          years_weight=float(number(item.get("yearsWeight"), 1.0)),
        )
      )
    periods.sort(key=lambda row: row.period)
    selected_raw = raw.get("selectedPeriod")
    selected_period = None
    if selected_raw not in (None, ""):
      try:
        selected_period = int(number(selected_raw))
      except Exception:
        selected_period = None
    return PathwayConfig(
      enabled=enabled,
      planning_mode="pathway" if enabled else "single_period",
      snapshot_mapping_mode=str(raw.get("snapshotMappingMode") or "explicit_period_column"),
      periods=periods,
      selected_period=selected_period,
    )
